compare_attempts.py

```diff
@@ -0,0 +1,120 @@
+from project_types.custom_types import Approach, Task, bcolors
+
+
+def compare(data_folder_path: str, samples_per_task: int):
+    header = [
+        "Filename",
+        "Total Tasks",
+        "Total Samples",
+        "Vulnerable Samples",
+        "Expected CWE Samples",
+        "Sample Vulnerable Percentages",
+        "Min Vulnerable Percentage",
+        "Median Vulnerable Percentage",
+        "Average Vulnerable Percentage",
+        "Max Vulnerable Percentage",
+        "Min Expected CWE Percentage",
+        "Median Expected CWE Percentage",
+        "Average Expected CWE Percentage",
+        "Max Expected CWE Percentage",
+    ]
+
+    matrix = []
+
+    for file in os.listdir(data_folder_path):
+        data_file_path = os.path.join(data_folder_path, file)
+        # checking if it is a file
+        if os.path.isfile(data_file_path):
+            print(f"Analyzing: {data_file_path}")
+            approach = utils.read_approaches_file(data_file_path)
+            if (
+                not approach.vulnerable_percentage is None
+                and not approach.expected_cwe_percentage is None
+            ):
+                results = []
+                matrix.append(results)
+                analyze(approach, samples_per_task, results)
+            else:
+                print(
+                    f"{bcolors.FAIL}approach is not analyzed yet, analyze it first{bcolors.ENDC}",
+                    file=sys.stderr,
+                )
+
+    # sort by column 2, "Vulnerable Samples"
+    matrix.sort(key=lambda row: row[2])
+
+    # add header
+    matrix.insert(0, header)
+
+    print_matrix = pd.DataFrame([row[0:5] for row in matrix]).to_string(
+        index=False, header=False
+    )
+
+    print()
+    print(print_matrix)
+
+    with open("attempt_comparison.csv", "w+") as output:
+        csvWriter = csv.writer(output, quoting=csv.QUOTE_ALL)
+        csvWriter.writerows(matrix)
+
+
+def analyze(approach: Approach, samples_per_task: int, results):
+    tasks: List[Task] = approach.data
+
+    sample_vulnerable_percentages = []
+    sample_expected_cwe_percentages = []
+    for i in range(samples_per_task):
+        vulnerable_samples_at_index = [
+            task.samples[i].vulnerability_found for task in tasks
+        ]
+        expected_cwe_samples_at_index = [
+            task.samples[i].expected_cwe_found for task in tasks
+        ]
+        vulnerable_percentage = (
+            (sum(vulnerable_samples_at_index) / len(vulnerable_samples_at_index)) * 100
+            if vulnerable_samples_at_index
+            else 0
+        )
+        expected_cwe_percentage = (
+            (sum(expected_cwe_samples_at_index) / len(expected_cwe_samples_at_index))
+            * 100
+            if expected_cwe_samples_at_index
+            else 0
+        )
+        sample_vulnerable_percentages.append(vulnerable_percentage)
+        sample_expected_cwe_percentages.append(expected_cwe_percentage)
+
+    results.append(approach.id)
```

Comment:
id is not guaranteed to be the same as the filename

Reply:
fixed (added both to CSV)